src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java

```diff
@@ -215,54 +219,120 @@ private void turnOffReadiness(Throwable exception){
             LOGGER.info("Created RabbitMQ connection {} [{}]", connection, connection.hashCode());
+            addShutdownListenerToConnection(this.connection);
+            addBlockedListenersToConnection(this.connection);
+            addRecoveryListenerToConnection(this.connection);
             metrics.getReadinessMonitor().enable();
             LOGGER.debug("Set RabbitMQ readiness to true");
         } catch (IOException | TimeoutException e) {
             metrics.getReadinessMonitor().disable();
             LOGGER.debug("Set RabbitMQ readiness to false. Can not create connection", e);
             throw new IllegalStateException("Failed to create RabbitMQ connection using configuration", e);
         }
+    }
 
-        this.connection.addBlockedListener(new BlockedListener() {
-            @Override
-            public void handleBlocked(String reason) {
-                LOGGER.warn("RabbitMQ blocked connection: {}", reason);
+    private void addShutdownListenerToChannel(Channel channel, Boolean withRecovery) {
+        channel.addShutdownListener(cause -> {
+            LOGGER.debug("Closing the channel: ", cause);
+            if (!cause.isHardError() && cause.getReference() instanceof Channel) {
+                Channel channelCause = (Channel) cause.getReference();
+                Method reason = cause.getReason();
+                if (reason instanceof AMQImpl.Channel.Close) {
+                    var castedReason = (AMQImpl.Channel.Close) reason;
+                    if (castedReason.getReplyCode() != 200) {
+                        StringBuilder errorBuilder = new StringBuilder("RabbitMQ soft error occupied: ");
+                        castedReason.appendArgumentDebugStringTo(errorBuilder);
+                        errorBuilder.append(" on channel ");
+                        errorBuilder.append(channelCause);
+                        String errorString = errorBuilder.toString();
+                        LOGGER.warn(errorString);
+                        if (withRecovery && errorString.contains("PRECONDITION_FAILED")) {
+                            recoverSubscriptionsOfChannel(channel);
+                        }
+                    }
+                }
             }
+        });
+    }
 
-            @Override
-            public void handleUnblocked() {
-                LOGGER.warn("RabbitMQ unblocked connection");
+    private void recoverSubscriptionsOfChannel(Channel channel) {
+        channelChecker.execute(() -> {
+            try {
+                var pinToChannelHolderOptional =
+                        channelsByPin
+                                .entrySet()
+                                .stream()
+                                .filter(entry -> Objects.nonNull(entry.getValue().channel) && channel.getChannelNumber() == entry.getValue().channel.getChannelNumber())
+                                .findAny();
+                if (pinToChannelHolderOptional.isPresent()) {
```

Comment:
Can we use `ifPresent` method here instead? Using an optional like this is no difference from getting the value with `orElse(null)` and checking that value for `null`. If we can't probably it is better to use `orElse(null)` method